Add relay set

andresilva opened the pull request:
Merge after #7.

- Refactored base behavior of `OwnedSet` into abstract `BaseOwnedSet`
- Added relay set contracts (`OuterSet`, `InnerSet`)
- Added `InnerOwnedSet` contract

andresilva commented:
Rebased over master, ready for review.

tomusdrw commented on contracts/OwnedSet.sol:
```diff
@@ -12,14 +12,15 @@
 
-// Owner can add or remove validators.
-contract OwnedSet is Owned, ValidatorSet {
+contract BaseOwnedSet is Owned, ValidatorSet {
```
Should we give some general structure to `contracts` directory?
Like:
```
`- contracts
   `- base/interfaces
   `- deployed (or just top-level)
   `- test
```
So that we have a clear view of what contracts should be deployed and which shouldn't?

This would imply splitting `OwnedSet.sol` into two files.

tomusdrw commented on contracts/RelaySet.sol:
```diff
@@ -0,0 +1,143 @@
+import "./ValidatorSet.sol";
+
+
+contract OuterSet is Owned, ValidatorSet {
```
Yeah, let's have a directory `ReleySet` and split the two into separate files.

tomusdrw commented on contracts/InnerOwnedSet.sol:
```diff
@@ -0,0 +1,75 @@
+
+
+contract InnerOwnedSet is Owned, InnerSet, BaseOwnedSet {
+	constructor(address _outerSet, address[] _initial) BaseOwnedSet(_initial)
```
Should the super constructor go to it's own line?

tomusdrw commented on contracts/RelaySet.sol:
```diff
@@ -0,0 +1,143 @@
+	// System address, used by the block sealer.
+	address public systemAddress;
+	// Address of the inner validator set contract
+	InnerSet public innerSet;
```
Maybe let's avoid generic types here and rather use `InnerOwnedSet` directly without any interface.

tomusdrw commented on contracts/InnerOwnedSet.sol:
```diff
@@ -0,0 +1,75 @@
+	constructor(address _outerSet, address[] _initial) BaseOwnedSet(_initial)
+		public
+	{
+		outerSet = OuterSet(_outerSet);
```
Should it invoke `setOuter` or maybe it should be left uninitialized until the owner calls `setOuter`?

tomusdrw commented on contracts/OwnedSet.sol:
```diff
@@ -150,24 +122,21 @@ contract OwnedSet is Owned, ValidatorSet {
 
-	// Report that a validator has misbehaved maliciously.
-	function reportMalicious(address _validator, uint _blockNumber, bytes _proof)
+	function reportBenign(address _validator, uint256 _blockNumber)
```
Should this be moved to `OwnedSet`? It seems that this contract should not expose those functions since it's used only to inherit from. I'd ditch implementing `ValidatorSet` interface for `BaseOwnedSet` and rather implement it lower in the hierarchy.

andresilva commented:
I've made some changes according to @tomusdrw's comments:

- Created folder structure for contracts
- Renamed `OuterSet` to `RelaySet` and removed `InnerSet` interface. Renamed `InnerOwnedSet` to `RelayedOwnedSet`
- Removed finality tracking from `RelaySet`, now it's a dumb proxy that relays requests
- Finality tracking was added to `BaseOwnedSet` and is shared for both implementations
- `BaseOwnedSet` no longer implements the `ValidatorSet` interface
- Rename base implementation methods to start with `base` prefix to make it more explicit

I will update this PR with better docs and some changes in tests.

andresilva commented:
I added a diagram to the README, the link should work when this is merged (it points to master).
